### ISSUE TITLE
Possible to keep this purely Go and cross-platform?

### ISSUE THREAD

**Issue by jbuberel** [EricLagerg/gonix#35]
Just tried to download and install on Windows (Win8.1) with Go 1.4.2, and see this:

```
$ make
go build  -o build/base64 base64.go
go build  -o build/basename basename.go
go build  -o build/cat cat.go
go build  -o build/cp cp.go
go build  -o build/dirname dirname.go
go build  -o build/echo echo.go
go build  -o build/false false.go
go build  -o build/head head.go
go build  -o build/ls ls.go
# command-line-arguments
exec: "gcc": executable file not found in %PATH%
make: *** [build/ls] Error 2
```

Looking at the source of `ls`, I see it relying on links to C structures. Is there an approach that would not require dropping into C and thus a full GCC build environment?

**Comment by mibk** [comment]
I agree. I cannot build now even on my machine runnig Arch Linux.
```
go build  -o build/ls ls.go
# command-line-arguments
./ls.go:109: cannot use C.gid_t(gid) (type C.gid_t) as type C.__gid_t in argument to _Cfunc_ge
tgrgid_r
./ls.go:134: invalid operation: stat.Nlink > mostLinksToFile (mismatched types uint64 and uint
16)
./ls.go:135: cannot use stat.Nlink (type uint64) as type uint16 in assignment
Makefile:19: recipe for target 'build/ls' failed
make: *** [build/ls] Error 2
```

I propose to make this only a pure Go project and ignore the stuff that requires additional "hacks" to run. I think it is wonderful job from @jutaz to get working even group id (as I think is the purpose of the dependency on C), but I'd like it to be **simple**.

**Comment by jbuberel** [comment]
If cross-platform isn't possible, remaining purely in Go would at least make it much easier to build on Linux-like operating systems.

Side note: I've been looking for a cross-platform approach to obtaining the dimensions of the current terminal. Looks like a harder problem than I had hoped.

**Comment by polegone** [comment]
ls doesn't compile for me on Deban either.

    go build  -o build/ls ls.go
    # command-line-arguments
    ls.go:109[/tmp/go-build548865991/command-line-arguments/_obj/ls.cgo1.go:154]: cannot use _Ctype_gid_t(gid) (type _Ctype_gid_t) as type _Ctype___gid_t in function argument
    ls.go:134[/tmp/go-build548865991/command-line-arguments/_obj/ls.cgo1.go:184]: invalid operation: stat.Nlink > mostLinksToFile (mismatched types uint32 and uint16)
    ls.go:135[/tmp/go-build548865991/command-line-arguments/_obj/ls.cgo1.go:185]: cannot use stat.Nlink (type uint32) as type uint16 in assignment
    make: *** [build/ls] Error 2

**Comment by vanhalt** [comment]
That `ls` implementation is a pro level code. I can't actually spot what is being used from C... Maybe if we spot that would be able to replace? Kinda lost there... hehe

**Comment by jutaz** [comment]
Oh, that is my bad that I did not test that on all *nix environments, sorry for that.

the problem here is that, for example Windows has a completely different approach to `ls` command, which is very very platform dependent, because there are syscalls involved. And windows, for example does not actually have `ls` implementation, so we would have to make some specific syscalls to get file owner, file owner group.

Having it "pure go", is not really that simple, as Go itself uses the same `C` API\`s you see in `ls.go` ([example](http://golang.org/src/os/user/lookup_unix.go)), and for example use lookup API is not available on windows, so I do not really see how we could implement the same `ls` for windows. We could omit some fields from `ls -la`, but then it would not be the same *nix `ls`.

And for *nix compile problems - those can be fixed quite easily, just need to  wrap some types a bit differently.

TL;DR - Windows implementation of `ls` would require not to have few syscalls and thus output would not be the same as on *nix. *nix implementations needs more work, but currently there is no way to make it pure Go.

Over Today and Tomorrow I will make some PR\`s that will hopefully clear the confusion in code, as @vanhalt pointed out, and will fix build fox *nix systems.

And what do we want to do with Windows? We could implement `ls` as `dir` is implemented in windows.

**Comment by jutaz** [comment]
Also, I will do much better job at having better cross platform implementation of getting term width. That requires more C API\`s to be called, but it should work much better.

+, if on installed GCC on windows, it should work too

**Comment by jbuberel** [comment]
Before we go down that path (no pun intended), I would like to hear from @polegone about the goals for this project.

Is the goal to (a) provide a replacement set of the standard command line unix tools that work across all of the platforms on which Go is available, or (b) provide a replacement set of standard unix tools on Linux?

If the answer is (b) and this project will only target Linux, my follow-up question would then be: Why do it at all?

**Comment by jutaz** [comment]
Right, completely agree with @jbuberel here, we should know what is the end goal here.

The funny thing is though, that most of these low level things are implemented as syscalls in Go, so to be truly cross platform we should either:
1. Wait for Go to implement all required syscalls in all platforms
2. Ditch some low level features like Getting group names in `ls`, etc.

**Comment by jbuberel** [comment]
<nod> If cross platform is the goal, then we'll likely have to give up on "complete compatibility" and instead shoot for "pretty good compatibility".

**Comment by mibk** [comment]
@jbuberel :+1:

**Comment by polegone** [comment]
I say we get it to work on all *nix platforms first (Linux/BSD/OS X), and decide what to do about Windows later.

I decided that we'll soon be merging into @EricLagerg's go-coreutils, and while he is aiming for cross-platformity, the commands that depend on users and groups (uname, chown, logname) do not work on Windows according to his table.

**Comment by mibk** [comment]
I think we can just not compile *groups* for Windows using build constraints https://golang.org/pkg/go/build/#hdr-Build_Constraints

**Comment by jutaz** [comment]
Yea, we can still be cross-platform, just _some_ things will be missing from windows

**Comment by jbuberel** [comment]
That seems reasonable to me. 

Is there a patter/approach you'd like to take on how to separate the CLI interface from the platform-specific implementations? Are there any good examples from the Go standard library that we should be following?

**Comment by jutaz** [comment]
What we can do, is to name files accordingly:

`ls.go` - stub
`ls_linux.go` - with linux syscalls
`ls_darwin.go` - with OS X specific code
`ls_windows.go` - with limited implementation.

This naming should be picked up by compiler and compiled only for appropriate platform

**Comment by vanhalt** [comment]
Yeah... actually is what @EricLagerg is doing, check his [wc](https://github.com/EricLagerg/go-coreutils/tree/master/wc) implementation.

**Comment by EricLagerg** [comment]
The C can actually be dropped if you use this patch: https://codereview.appspot.com/101310044/

Go 1.5 *should* have this implemented... I think.

Also, +1 on the `ls_*.go` conventions.

**Comment by jutaz** [comment]
Well, that uses the same API's I use, so I see no dropping of C here 😉

But the naming should fix this C issue anyway

**Comment by EricLagerg** [comment]
That's true. Having it integrated in the stdlib makes it that much prettier (not to mention how compilation will be much faster since each compilation of `gonix` will just have to use the `.a` files instead of compiling the C!)

**Comment by 0xCMP** [comment]
And technically we wouldn’t be using C.


-- cmp
